### ISSUE TITLE
`count (dirs)` doesn't work as expected

### ISSUE THREAD

**Issue by andmis** [fish-shell/fish-shell#9359]
IMO this is confusing behavior:

```
andrey@oai-andrey /tmp> fish --version
fish, version 3.5.1
andrey@oai-andrey /tmp> uname -a
Darwin oai-andrey 22.1.0 Darwin Kernel Version 22.1.0: Sun Oct  9 20:15:09 PDT 2022; root:xnu-8792.41.9~2/RELEASE_ARM64_T6000 arm64
andrey@oai-andrey /tmp> echo $TERM
xterm-256color
andrey@oai-andrey /tmp> pushd /var
andrey@oai-andrey /var> pushd /usr
andrey@oai-andrey /usr> dirs
/usr /var /tmp
andrey@oai-andrey /usr> count (dirs)
1
andrey@oai-andrey /usr>
```

I would expect `count (dirs)` to give `3` in this case. (I do understand why it doesn't, technically.)

---

The natural workaround is to `string split`, but this doesn't work well with directories with spaces:

```
andrey@oai-andrey /usr> cd /tmp
andrey@oai-andrey /tmp> mkdir dir\ with\ spaces
andrey@oai-andrey /tmp> pushd dir\ with\ spaces/
andrey@oai-andrey /t/dir with spaces> dirs
/tmp/dir with spaces /tmp /var /tmp
andrey@oai-andrey /t/dir with spaces> count (dirs)
1
andrey@oai-andrey /t/dir with spaces> count (string split ' ' (dirs))
6
andrey@oai-andrey /t/dir with spaces>
```

---

Making `dirs` split with newlines would solve the issue. Dunno how breaking this change would be. We could also add a command line argument, but that's a bit annoying.

---

Also, `man dirs` for me says:

```
       The following options are available:

       -c:    Clear the directory stack instead of printing it.

       -h or --help
              Displays help about using this command.

       dirs does not accept any arguments.
```

It looks like `dirs does not accept any arguments` is wrong.

**Comment by mqudsi** [comment]
You are right and this is very unfortunate. As a first-party script, there's no reason why `dirs` should emit info in a way not natively ingestible by other fish scripts. We could do a GitHub code search survey to see what the impact of changing the default format would be, otherwise we can add an option to split with new lines.

**Comment by floam** [comment]
My opinion is, let's just change it to be newline-delimited.

**Comment by faho** [comment]
My gut tells me `dirs` probably isn't used a lot in scripts.

It's a thing to handle the directory stack, and when you're writing a script that seems like weird baggage.

But I'd like to have a bit of a rethink of all our awkward directory stack related functions in general?

We have:

1. `dirs`, which prints $dirstack on one line and can clear it
2. `dirh` which prints $dirnext and $dirprev on separate lines, numbered, no ability to clear
3. `cd`, which adds to $dirnext/$dirprev
4. `prevd`/`next`, which move through $dirnext/$dirprev
5. `cdh`, which allows picking from $dirprev/$dirnext via a menu
6. `pushd`/`popd`, which add/pop from $dirstack

Those are a lot of functions for something extremely simple and related, and I'd argue we could do for some consolidation here, or possibly some better naming.

In this specific case, what you can already do is `count $dirstack`.

>It looks like dirs does not accept any arguments is wrong.

"arguments" in this case refers to *positional* arguments. Anything that's not seen as an option is an error here.

**Comment by mqudsi** [comment]
I'd be onboard with improving the current api surface for the directory stack, and agree it's unfortunate that there are too many separate commands to do everything. I feel like `dirs`, `dirh`, `pushd`, and `popd` should probably all have been subcommands or options to one builtin. `cd` is obviously untouchable, and `cdh` I have no opinions about (it could very well have been `cd -h`). 

For `prevd` and `nextd`, it's a bit trickier. I don't know about everyone else, but for me their names are too long for what they accomplish. `prevd` is in lieu of `cd -` , `nextd` doesn't have a direct parallel, but in all cases, I have these aliased to `pd` and `nd`. 

What we *could* do is unify everything under `cd` and `dirs`, and put default aliases for the old names. That preserves backwards compatibility but leaves it possible for someone to replace the the alias without losing access to the functionality. I'm not entirely sure if the effort is worth it, but it does clean up at least the hierarchy of commands and the docs, making it easier to discover everything to do with dirs in one place.

**Comment by andmis** [comment]
I do use `pushd` and `popd` sometimes, but it occurs to me that I could just as well call `fish` to spawn a subshell. What are the major downsides to using `fish`/CTRL-D instead of `pushd/popd`?

**Comment by mqudsi** [comment]
> you lose prompt edits in your session coming from a Python virtualenv

I think `history merge` does what you want?

**Comment by andmis** [comment]
```
~$ cd tmp
~/tmp$ python3.9 -m venv venv
~/tmp$ . venv/bin/activate.fish
(venv) ~/tmp$ fish
~/tmp$ history merge
~/tmp$ 
```

Sorry, just to be clear: Python venvs edit your prompt, like shown above. I don't think `history merge` does anything to address this.
